bindings/resolver.go

```diff
@@ -0,0 +1,253 @@
+		return Binding{}, err
+	}
+	if len(binds) != 1 {
+		return Binding{}, fmt.Errorf("found %d bindings for type '%s' and provider '%s' but expected exactly 1", len(binds), typ, provider)
+	}
+	return binds[0], nil
+}
+
+func loadBindings(bindingRoot string) ([]Binding, error) {
+	files, err := ioutil.ReadDir(bindingRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	var bindings []Binding
+	for _, file := range files {
+		isLegacy, err := isLegacyBinding(bindingRoot, file.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		var binding Binding
+		if isLegacy {
+			binding, err = loadLegacyBinding(bindingRoot, file.Name())
+		} else {
+			binding, err = loadBinding(bindingRoot, file.Name())
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading binding '%s'", file.Name())
+		}
+		bindings = append(bindings, binding)
+	}
+	return bindings, nil
+}
+
+// According to the legacy spec (https://github.com/buildpacks/spec/blob/main/extensions/bindings.md), a legacy binding
+// has a `metadata` directory within the binding path.
+func isLegacyBinding(bindingRoot string, name string) (bool, error) {
+	info, err := os.Stat(filepath.Join(bindingRoot, name, "metadata"))
+	if err == nil {
+		return info.IsDir(), nil
+	} else if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+// See: https://github.com/k8s-service-bindings/spec#workload-projection
+func loadBinding(bindingRoot string, name string) (Binding, error) {
+	binding := Binding{
+		Name:   name,
+		Path:   filepath.Join(bindingRoot, name),
+		Secret: map[string]SecretReader{},
+	}
+
+	secrets, err := loadSecrets(filepath.Join(binding.Path))
+	if err != nil {
+		return Binding{}, err
+	}
+
+	typ, ok := secrets["type"]
+	if !ok {
+		return Binding{}, errors.New("missing 'type'")
+	}
+	binding.Type = typ.String()
+	delete(secrets, "type")
+
+	provider, ok := secrets["provider"]
+	if ok {
+		binding.Provider = provider.String()
+		delete(secrets, "provider")
+	}
+
+	binding.Secret = secrets
+
+	return binding, nil
+}
+
+// See: https://github.com/buildpacks/spec/blob/main/extensions/bindings.md
+func loadLegacyBinding(bindingRoot string, name string) (Binding, error) {
+	binding := Binding{
+		Name:   name,
+		Path:   filepath.Join(bindingRoot, name),
+		Secret: map[string]SecretReader{},
+	}
+
+	metadata, err := loadSecrets(filepath.Join(binding.Path, "metadata"))
+	if err != nil {
+		return Binding{}, err
+	}
+
+	typ, ok := metadata["kind"]
+	if !ok {
+		return Binding{}, errors.New("missing 'kind'")
+	}
+	binding.Type = typ.String()
+	delete(metadata, "kind")
+
+	provider, ok := metadata["provider"]
+	if !ok {
+		return Binding{}, errors.New("missing 'provider'")
+	}
+	binding.Provider = provider.String()
+	delete(metadata, "provider")
+
+	binding.Secret = metadata
+
+	secrets, err := loadSecrets(filepath.Join(binding.Path, "secret"))
+	if err != nil && !os.IsNotExist(err) {
+		return Binding{}, err
+	}
+	if err == nil {
+		for k, v := range secrets {
```

Comment:
Technically, this means that any keys within `secret` will trump keys within `metadata`. I don't love it, but it's how existing buildpack implementations handle legacy bindings (via [libcnb](https://github.com/buildpacks/libcnb/blob/main/platform.go#L100)), so it's nothing new.

To handle it differently, there are a couple of options:
- Return two different binding types when resolving: `Binding` or `LegacyBinding` (meh)
- Open up `Binding` to separate `metadata` from `secret` into separate fields (results in unused field for newer binding type and could be confusing)

Overall, it's such an edge case that I don't think it's worth those refactors to address, as it's only for the legacy binding format. No ambiguity for the new format.

Thoughts?

Reply:
I'm not sure how often the collision will occur, but I'm open to the second option you've proposed.  I would rather that the case for the k8s-based binding spec was adhered more strictly (no empty fields) than the legacy-based binding  since that's the direction we are moving in in the long run. 

I would also be okay with leaving the code as is in this case, and then revisiting this if we get user feedback about this being an issue.

Reply:
OK, I'm also in favor of leaving as is. I think if this were a big risk, it would have turned up by now with all the Java buildpacks' implementations (which use `libcnb` under the hood).